lib/bashcov/xtrace.rb

```diff
@@ -1,29 +1,67 @@
+# frozen_string_literal: true
+
+require "pathname"
+require "securerandom"
+
+require "bashcov/errors"
+
 module Bashcov
   # This class manages +xtrace+ output.
   #
   # @see Runner
   class Xtrace
-    # Prefix used for PS4.
-    # @note The first caracter ('+') will be repeated to indicate the nesting
-    #   level.
-    PREFIX = "+BASHCOV> "
-
-    # [String] +PS4+ variable used for xtrace output
-    # @see http://www.gnu.org/software/bash/manual/bashref.html#index-PS4
-    # @see http://stackoverflow.com/questions/59895/can-a-bash-script-tell-what-directory-its-stored-in
-    # @note We use a forward slash as delimiter since it's the only forbidden
-    #   character in filenames on Unix and Windows.
-    GET_ABS_DIR = "$(cd $(dirname ${BASH_SOURCE[0]}); pwd)"
-    GET_BASE = "$(basename ${BASH_SOURCE[0]})"
-    PS4 = %(#{PREFIX}#{GET_ABS_DIR}/#{GET_BASE}/${LINENO}: )
-
-    # Regexp to match xtrace elements.
-    LINE_REGEXP = %r{\A#{Regexp.escape(PREFIX[0])}+#{PREFIX[1..-1]}(?<filename>.+)\/(?<lineno>\d+): }
+    # [String] Character that will be used to indicate the nesting level of
+    #   +xtrace+d instructions
+    DEPTH_CHAR = "+".freeze
+
+    # [String] Prefix used in +PS4+ to identify relevant output
+    PREFIX = "BASHCOV>".freeze
+
+    # [Array<String>] A collection of Bash internal variables to expand in the
+    #   {PS4}
+    FIELDS = %w(${LINENO} ${BASH_SOURCE} ${PWD} ${OLDPWD}).freeze
+
+    class << self
+      attr_writer :delim, :ps4
+
+      # [String] A randomly-generated UUID or the ASCII RS (record separator)
+      #   character, depending on whether the current Bash suffers from the
+      #   truncated +PS4+ bug.  Used for delimiting the fields of the +PS4+.
+      def delim
+        @delim ||= (Bashcov.truncated_ps4? && !Bashcov.use_trap) ? "\x1E" : SecureRandom.uuid
```

Comment:
Whoops, good eye -- that's a leftover from an experiment using the `DEBUG` trap to capture coverage data.